std/exception.d

```diff
@@ -2148,3 +2148,65 @@ pure nothrow @safe unittest
 
     auto infSlice = infinite[0 .. $]; // this would throw otherwise
 }
+
+
+/++
+    Convenience mixin for trivially sub-classing exceptions
+
+    Even trivially sub-classing an exception involves writing boilerplate code
+    for the constructor to: 1) correctly pass in the source file and line number
+    the exception was thrown from; 2) be usable with $(LREF enforce) which
+    expects exception constructors to take arguments in a fixed order. This
+    mixin provides that boilerplate code.
+
+    Note however that you need to mark the $(B mixin) line with at least a
+    minimal (i.e. just $(B ///) DDoc comment if you want the mixed-in
+    constructors to be documented in the newly created Exception subclass.
+ +/
+mixin template basicExceptionCtors()
+{
+    /++
+        Params:
+            msg  = The message for the exception.
+            file = The file where the exception occurred.
+            line = The line number where the exception occurred.
+            next = The previous exception in the chain of exceptions, if any.
+    +/
+    this(string msg, string file = __FILE__, size_t line = __LINE__,
+         Throwable next = null) @safe pure nothrow
+    {
+        super(msg, file, line, next);
+    }
+
+    /++
+        Params:
+            msg  = The message for the exception.
+            next = The previous exception in the chain of exceptions.
+            file = The file where the exception occurred.
+            line = The line number where the exception occurred.
+    +/
+    this(string msg, Throwable next, string file = __FILE__,
+         size_t line = __LINE__) @safe pure nothrow
+    {
+        super(msg, file, line, next);
+    }
+}
+
+///
+pure nothrow @safe unittest
+{
+    class MeaCulpa: Exception
+    {
+        ///
+        mixin basicExceptionCtors;
+    }
+
+    try
+        throw new MeaCulpa("test");
+    catch (Exception e) // Bugzilla 15467
```

Comment:
There is no requirement that you have only one `unittest` block per function or type. So, you can remove `nothrow` from the ddoc-ed `unttest` block that you have now so that you can catch `MeaCulpa`. Then create another `unittest` block after it which is marked with attributes like `nothrow` to enforce that any code within them has those attributes and construct exceptions in there to test that the mixed in constructors have the appropriate attributes. While you're at it, you can test the constructors more thoroughly than makes sense in the ddoc example to make sure that the various ways of calling them all work. Right now, you're only testing one of the constructors and only with its default arguments. e.g.

    pure nothrow @safe unittest
    {
        class TestException : Exception { mixin basicExceptionCtors; }
        auto e = new Exception("!!!");
        auto te1 = new TestException("message", "file", 42, e);
        assert(te1.message == "message");
        assert(te1.file == "file");
        assert(te1.line == 42);
        assert(te1.next is e);

        auto te2 = new TestException(e, "message", "file", 42);
        assert(te2.message == "message");
        assert(te2.file == "file");
        assert(te2.line == 42);
        assert(te2.next is e);

        auto te3 = new TestException("foo");
        assert(te3.message == "foo");
        assert(te3.file == __FILE__);
        assert(te3.line == __LINE__ - 3);
        assert(te3.next is null);

        auto te4 = new TestException(e, "foo");
        assert(te4.message == "foo");
        assert(te4.file == __FILE__);
        assert(te4.line == __LINE__ - 3);
        assert(te4.next is e);
    }

So, you can leave the current `unittest`block mostly as-is (just remove the `nothrow` and change it so that it catches `MeaCulpa`), and then the second `unittest` block can test the rest of what should be tested but doesn't need to be in the documentation.